apps/api/src/app/rate-limiting/usecases/get-default-api-rate-limits/get-default-api-rate-limits.usecase.ts

```diff
@@ -0,0 +1,50 @@
+import { Injectable } from '@nestjs/common';
+import {
+  ApiRateLimitCategoryTypeEnum,
+  ApiRateLimitEnvVarFormat,
+  ApiServiceLevelTypeEnum,
+  DEFAULT_API_RATE_LIMITS,
+  IServiceApiRateLimits,
+} from '@novu/shared';
+
+@Injectable()
+export class GetDefaultApiRateLimits {
+  public defaultApiRateLimits: IServiceApiRateLimits;
+
+  constructor() {
+    this.loadApiRateLimits();
+  }
+
+  public loadApiRateLimits(): void {
+    this.defaultApiRateLimits = this.createDefaultApiRateLimits();
+  }
+
+  private createDefaultApiRateLimits(): IServiceApiRateLimits {
+    const mergedApiRateLimits: IServiceApiRateLimits = { ...DEFAULT_API_RATE_LIMITS };
+
+    // Read process environment only once for performance
+    const processEnv = process.env;
+
+    Object.values(ApiServiceLevelTypeEnum).forEach((apiServiceLevel) => {
+      Object.values(ApiRateLimitCategoryTypeEnum).forEach((apiRateLimitCategory) => {
+        const envVarName = this.getEnvVarName(apiServiceLevel, apiRateLimitCategory);
+        const envVarValue = processEnv[envVarName];
+
+        if (envVarValue) {
+          mergedApiRateLimits[apiServiceLevel][apiRateLimitCategory] = Number(envVarValue);
+        }
+      });
+    });
```

Comment:
You're probably right in that it's a premature optimisation.

My curiosity was piqued to run a performance test on the impact of reading from `process.env`, here it is:
<details><summary>Result</summary>
<p>
The script runs an inner process.env read loop 100 times, then runs each of the inner loops experiments 100,000 times to eliminate local machine usage variance.

I have then run the script from the console a further 10 times to further reduce local machine variance and reduce local machine caching impact.

</p>
<details><summary>Script</summary>
```javascript
process.on('unhandledRejection', (err)=>{
 console.error(err);
})

const INNER_LOOP_ITERATIONS = 100;
const TEST_ITERATIONS = 100000;

function benchmark(func, iterations) {
  let timeTotal = 0;
  for (let i = 0; i < iterations; i++) {
    const start = process.hrtime();
    func();
    const diff = process.hrtime(start);
    const timeInSeconds = diff[0] + diff[1] / 1e9;
    timeTotal += timeInSeconds;
  }
  const average = timeTotal / iterations;
  return average;
}

function readOnce() {
  const processEnv = process.env;
  const func = () => {
    const t = processEnv['TZ'];
  };
  const time = benchmark(func, INNER_LOOP_ITERATIONS);
  return time;
}

function readEvery() {
  const func = () => {
    const tz = process.env['TZ'];
  };
  const time = benchmark(func, INNER_LOOP_ITERATIONS);
  return time;
}

function main() {
  let timeEveryTotal = 0;
  let timeOnceTotal = 0;

  // Run the test twice in different order to avoid any caching
  for (let i = 0; i < TEST_ITERATIONS; i++) {
    timeEveryTotal += readEvery();
    timeOnceTotal += readOnce();
  }

  for (let i = 0; i < TEST_ITERATIONS; i++) {
    timeOnceTotal += readOnce();
    timeEveryTotal += readEvery();
  }

  const timeOnce = timeOnceTotal / (TEST_ITERATIONS * 2);
  const timeEvery = timeEveryTotal / (TEST_ITERATIONS * 2);
  console.log(`Once\t Time:\t ${timeOnce.toPrecision(3)}s`);
  console.log(`Every\t Time:\t ${timeEvery.toPrecision(3)}s`);
  console.log(`Ratio\t Perc:\t ${(timeOnce/timeEvery*100).toPrecision(3)}%`)
  console.log(`Bonus\t Perc:\t ${((1 - timeOnce/timeEvery)*100).toPrecision(3)}%`)
}

main()
```
</details> 
<details><summary>Script run results</summary>

```bash
rifont at Richards-MacBook-Pro in ~/Downloads                        15:29
> node env-test.js
Once	 Time:	 1.65e-7s
Every	 Time:	 1.65e-7s
Ratio	 Perc:	 100%
Bonus	 Perc:	 -0.301%
rifont at Richards-MacBook-Pro in ~/Downloads                        15:29
> node env-test.js
Once	 Time:	 1.65e-7s
Every	 Time:	 1.66e-7s
Ratio	 Perc:	 99.4%
Bonus	 Perc:	 0.598%
rifont at Richards-MacBook-Pro in ~/Downloads                        15:29
> node env-test.js
Once	 Time:	 1.60e-7s
Every	 Time:	 1.65e-7s
Ratio	 Perc:	 97.1%
Bonus	 Perc:	 2.91%
rifont at Richards-MacBook-Pro in ~/Downloads                        15:29
> node env-test.js
Once	 Time:	 1.74e-7s
Every	 Time:	 1.88e-7s
Ratio	 Perc:	 92.6%
Bonus	 Perc:	 7.43%
rifont at Richards-MacBook-Pro in ~/Downloads                        15:30
> node env-test.js
Once	 Time:	 1.69e-7s
Every	 Time:	 1.68e-7s
Ratio	 Perc:	 101%
Bonus	 Perc:	 -0.551%
rifont at Richards-MacBook-Pro in ~/Downloads                        15:30
> node env-test.js
Once	 Time:	 1.61e-7s
Every	 Time:	 1.64e-7s
Ratio	 Perc:	 98.2%
Bonus	 Perc:	 1.79%
rifont at Richards-MacBook-Pro in ~/Downloads                        15:30
> node env-test.js
Once	 Time:	 1.64e-7s
Every	 Time:	 1.66e-7s
Ratio	 Perc:	 99.1%
Bonus	 Perc:	 0.912%
rifont at Richards-MacBook-Pro in ~/Downloads                        15:30
> node env-test.js
Once	 Time:	 1.66e-7s
Every	 Time:	 1.70e-7s
Ratio	 Perc:	 97.7%
Bonus	 Perc:	 2.33%
rifont at Richards-MacBook-Pro in ~/Downloads                        15:31
> node env-test.js
Once	 Time:	 1.64e-7s
Every	 Time:	 1.66e-7s
Ratio	 Perc:	 98.8%
Bonus	 Perc:	 1.19%
rifont at Richards-MacBook-Pro in ~/Downloads                        15:31
> node env-test.js
Once	 Time:	 1.60e-7s
Every	 Time:	 1.66e-7s
Ratio	 Perc:	 96.3%
Bonus	 Perc:	 3.72%
```
</details> 
</details> 

Average improvement for reading once versus reading every iteration was 2.002%, with an average time to read the process.env of 1.648e-7 seconds, or 0.1648 microseconds. With an average API request duration of 0.3 seconds, this bonus accounts for a 0.0000549333% speed increase. Not really worth it!

Reply:
Incredible work! Thanks for checking that.
Anyway that is in one single place and we have a lot of places where the env variables are being read from process.env. Maybe we could think of having a single reading point and reused everywhere of the platform, but just being careful that we can override them for the testing needs.
Hopefully Node.js would implement any caching system for the process.env in the future.